File: wallet-application/src/main/java/com/example/wallet/application/command/PlaceOrderCommand.java
package com.example.wallet.application.command;

import com.example.wallet.domain.order.OrderItem;
import com.example.wallet.domain.shared.Money;

import java.math.BigDecimal;
import java.util.Currency;
import java.util.List;

public record PlaceOrderCommand(
        String idempotencyKey,
        String buyerId,
        Currency currency,
        List<OrderLine> lines
) {
    public record OrderLine(String sku, int quantity, BigDecimal unitPrice) {}

    public List<OrderItem> toOrderItems() {
        return lines.stream()
                .map(l -> new OrderItem(l.sku(), l.quantity(), Money.of(l.unitPrice(), currency)))
                .toList();
    }
}

File: billing-application/src/main/kotlin/com/example/billing/application/command/PlaceOrderCommand.kt
package com.example.billing.application.command

import com.example.billing.domain.order.OrderItem
import com.example.billing.domain.shared.Money
import java.math.BigDecimal
import java.util.Currency

@JvmRecord
data class PlaceOrderCommand(
    val idempotencyKey: String,
    val buyerId: String,
    val currency: Currency,
    val lines: List<OrderLine>,
) {

    fun toOrderItems(): List<OrderItem> =
        lines.map { OrderItem(it.sku, it.quantity, Money.of(it.unitPrice, currency)) }

    @JvmRecord
    data class OrderLine(val sku: String, val quantity: Int, val unitPrice: BigDecimal)
}
